mof_array/pmf/process_mass_data.py

```diff
@@ -318,55 +318,64 @@ def create_bins(gases, num_bins, mof_list, element_pmf_results):
-        for row in array_pmf_results:
-             for i in range(1, len(create_bins_results)):
-                if ( float(row[gas_name]) >= create_bins_results[i - 1][gas_name] and
-                     float(row[gas_name]) < create_bins_results[i][gas_name]
-                   ):
-                    row.update({'%s bin' % gas_name: create_bins_results[i - 1][gas_name]})
+        # Current method likely inefficient, but it works. Can revisit later.
+        for row in all_array_pmf_results:
+             for i in range(1, len(bins)):
+                 lower_bin = bins[i-1][gas]
+                 upper_bin = bins[i][gas]
+                 gas_comp = float(row[gas])
+                 if gas_comp >= lower_bin and gas_comp < upper_bin:
+                     row.update({'%s bin' % gas: lower_bin})
 
         # Loops through all of the bins and takes sum over all pmfs in that bin.
-        binned_probability_temporary = []
-        for b in create_bins_results[0:len(create_bins_results)-1]:
-            temp_array_pmf = {'%s' % ' '.join(array): [] for array in list_of_arrays}
-            for line in array_pmf_results:
-                # Checks that the gas' mole frac matches the current bin
-                if b[gas_name] == line['%s bin' % gas_name]:
-                    # For each array, assigns the pmfs to their corresponding key
-                    for array in list_of_arrays:
-                        temp_pmf_list = temp_array_pmf['%s' % ' '.join(array)]
-                        temp_pmf_list.append(line['%s' % ' '.join(array)])
-                        temp_array_pmf['%s' % ' '.join(array)] = temp_pmf_list
-
-            # Updates pmfs for each array for current bin, summing over all pmfs
-            bin_temporary = {'%s bin' % gas_name : b[gas_name]}
-            for array in list_of_arrays:
-                if temp_array_pmf['%s' % ' '.join(array)] == []:
-                    bin_temporary.update({'%s' % ' '.join(array): 0})
+        all_bins_temp_sum = []
+        all_bins_temp_max = []
+        array_names = ['%s' % ' '.join(array) for array in list_of_arrays]
+        for bin in bins[0:len(bins)-1]:
+            array_pmfs_temp = {array: [] for array in array_names}
+            for row in all_array_pmf_results:
+                if bin[gas] == row['%s bin' % gas]:
+                    for array in array_names:
+                        array_pmfs_temp[array].append(row[array])
+
+            # Updates pmfs for each array for current bin
+            # Many methods here of handling multiple data points.
+            # Currently, can sum all pmfs, or take the max value.
+            single_bin_temp = {'%s bin' % gas : bin[gas]}
+            with_sum = copy.deepcopy(single_bin_temp)
+            with_max = copy.deepcopy(single_bin_temp)
+            for array in array_names:
+                if array_pmfs_temp[array] == []:
+                    with_sum.update({array : 0})
+                    with_max.update({array : 0})
```

Comment:
Can we just use normal syntax: `with_sum[array] = 0` ?

Reply:
The temporary dictionary which is being updated, single_bin_temp, does not yet include the key for the array, so there would be a key error, hence the need for update. 

I know we talked about pre-allocating all these keys, or switching to an object for managing this data, but I was going to save that for a separate exercise / set of commits on a new feature branch following this PR.

Reply:
`dictionary[key] = value` doesn't require that the key exist in the dictionary before you use it. Typically you would only use update when you wanted to merge two dictionaries, or if you had a lot of key-value pairs you wanted to update in your dictionary, but if you only have one key-value pair to set, you typically use the standard `dictionary[key] = value` syntax. 

If you are going to rewrite this, I'm ok with you leaving it, as long as my above point is clear...?